httpcore/_utils.py

```diff
@@ -63,3 +64,20 @@ def origin_to_url_string(origin: Origin) -> str:
     scheme, host, explicit_port = origin
     port = f":{explicit_port}" if explicit_port != DEFAULT_PORTS[scheme] else ""
     return f"{scheme.decode('ascii')}://{host.decode('ascii')}{port}"
+
+
+def is_socket_at_eof(sock_fd: int) -> bool:
+    # Duplicate the socket from the file descriptor. (we do this so we always get a
+    # real `socket.socket` object, rather than a library-provided socket-like object,
+    # which may behave differently and/or may not have an implementation for `.recv()`.)
+    sock = socket.fromfd(sock_fd, socket.AF_INET, socket.SOCK_STREAM)
+    # Then put the copy into non-blocking mode. (We need this so that the `.recv()` call
+    # does not block.)
+    sock.setblocking(False)
+
+    try:
+        data = sock.recv(1, socket.MSG_PEEK)
```

Comment:
Hello,

I'm sorry for the question, do we really need to copy the socket? May be we can just try to
`data = original_socket.recv(1, socket. MSG_PEEK | socket. MSG_DONTWAIT)`

Reply:
Yes that's what I'd like to do, but as I've put in another comment in some cases this won't work.

Eg one weird case is when rubbing anyio in asyncio with uvloop: the socket object we get is a weird `PseudoSocket` instance, and when you try to call `.recv()` on it it raises an error saying it's not implemented. That's actually the only case I've seen of this for now, and maybe an okay strategy would be "recv(), and if it doesn't succeed copy the socket temporarily to get a real socket instance, and call recv() again". Then the "TCP only" limitation would only apply to the latter case, which should be a minority.

Reply:
What if you try:
```python
sock = socket(fileno=sock_fd)
data = sock.recv(1, socket. MSG_PEEK | socket. MSG_DONTWAIT)
```

According to the [official doc](https://docs.python.org/3/library/socket.html#socket.socket):
> If fileno is specified, the values for family, type, and proto are auto-detected from the specified file descriptor

and
> Unlike socket.fromfd(), fileno will return the same socket and not a duplicate

so you don't need to close it

Reply:
Wonderful. 💞

Reply:
Unfortunately according to the [python source](https://github.com/python/cpython/blob/deb016224cc506503fb05e821a60158c83918ed4/Lib/test/test_socket.py#L2986) `MSG_DONTWAIT` is for Linux only.

Reply:
(but probably we still need to close the `sock`)